### PR TITLE
Refactor media to use cloudinary

### PR DESCRIPTION
**Pull request by florianschmidt1994**
Use cloudinary and add endpoints for media handling, instead of relying on previous manuel handling.
This includes breaking changes in the API

We should test before merging:

- [ ] Does uploading work
- [ ] Does deletion work

### CODE REVIEW

**Comment by thepiwo** [comment]
there is still comments of how to handle sponsoring contracts, how should we proceed on this?

File: src/main/java/backend/model/challenges/Challenge.kt [hunk 1]
```diff
@@ -139,7 +138,7 @@ class Challenge : BasicEntity, Billable {
         this.team = team
         this.amount = amount
         this.description = description
-        this.contract = Media(DOCUMENT)
+        this.contract = null //TODO: how to handle contracts in future?
```
Comment (florianschmidt1994) [inline src/main/java/backend/model/challenges/Challenge.kt:142]: I'd say revert that change and keep type document? We want to handle it in the same way, we can still upload assets to cloudinary even if they are not images or videos?

File: src/main/java/backend/model/event/Team.kt [hunk 2]
```diff
@@ -52,8 +52,8 @@ class Team : BasicEntity {
     @OneToMany(cascade = arrayOf(ALL), mappedBy = "team", orphanRemoval = true)
     private var invitations: MutableList<Invitation> = ArrayList()
 
-    @OneToOne(cascade = arrayOf(ALL), orphanRemoval = true, fetch = LAZY)
-    lateinit var profilePic: Media
+    @ManyToOne(fetch = LAZY)
+    var profilePic: Media? = null
```
Comment (florianschmidt1994) [inline src/main/java/backend/model/event/Team.kt]: Why many to one? `profilePic 1-->1 team` i'd say. One profile pic belongs to one team, and each team has only one profile picture?

File: src/main/java/backend/model/posting/Posting.kt [hunk 3]
```diff
@@ -45,8 +45,8 @@ class Posting : BasicEntity {
     @ManyToOne(fetch = LAZY)
     var team: Team? = null
 
-    @OneToMany(cascade = arrayOf(CascadeType.ALL))
-    var media: MutableList<Media> = arrayListOf()
+    @ManyToOne(fetch = LAZY)
```
Comment (florianschmidt1994) [inline src/main/java/backend/model/posting/Posting.kt]: Why not one to one?

File: src/main/java/backend/model/sponsoring/Sponsoring.kt [hunk 4]
```diff
@@ -62,7 +61,7 @@ class Sponsoring : BasicEntity, Billable {
         this.team = team
         this.amountPerKm = amountPerKm
         this.limit = limit
-        this.contract = Media(DOCUMENT)
+        this.contract = null
```
Comment (florianschmidt1994) [inline src/main/java/backend/model/sponsoring/Sponsoring.kt:64]: Why set this to null?

File: src/main/java/backend/model/sponsoring/Sponsoring.kt [hunk 5]
```diff
@@ -72,7 +71,7 @@ class Sponsoring : BasicEntity, Billable {
         this.amountPerKm = amountPerKm
         this.limit = limit
         this.status = ACCEPTED
-        this.contract = Media(DOCUMENT)
+        this.contract = null
```
Comment (florianschmidt1994) [inline src/main/java/backend/model/sponsoring/Sponsoring.kt:74]: Why set this to null? We could still keep this with document as media type

File: src/main/java/backend/model/user/UserAccount.kt [hunk 6]
```diff
@@ -43,8 +43,8 @@ class UserAccount : BasicEntity, User {
     @Enumerated(EnumType.STRING)
     override var preferredLanguage: Language = Language.DE
 
-    @OneToOne(cascade = arrayOf(CascadeType.ALL), orphanRemoval = true)
-    override var profilePic: Media
+    @ManyToOne(fetch = FetchType.LAZY)
```
Comment (florianschmidt1994) [inline src/main/java/backend/model/user/UserAccount.kt]: Why many to one?

File: src/main/java/backend/util/GeoTools.kt [hunk 7]
```diff
@@ -41,12 +41,6 @@ fun distanceCoordsKM(from: Coord, to: Coord): Double {
     return EarthCalc.getVincentyDistance(fromPoint, toPoint) / 1000
 }
 
-fun distanceCoordsListKMfromStart(startingPoint: Coord, list: List<Coord>): Double {
-    val coordsList = arrayListOf(startingPoint)
-    coordsList.addAll(list)
-    return distanceCoordsListKM(coordsList)
-}
-
```
Comment (florianschmidt1994) [inline src/main/java/backend/util/GeoTools.kt]: Why are we removing this here? Does it belong to this PR?

File: src/main/java/backend/util/StringsHelper.kt [hunk 8]
```diff
@@ -1,20 +0,0 @@
...
-    subject = subject
-            .replace("Ä", "Ae").replace("Ü", "Ue").replace("Ö", "Oe")
-            .replace("ä", "ae").replace("ü", "ue").replace("ö", "oe").replace("ß", "ss")
-            .replace(Regex("[^A-Za-z0-9-]"), "")
-    if (subject.length > 140) {
-        return subject.substring(0, 140)
-    } else {
-        return subject
-    }
-}
```
Comment (florianschmidt1994) [inline src/main/java/backend/util/StringsHelper.kt]: Why are we removing this here? Does it belong to this PR?

File: src/test/java/backend/Integration/TestMessagingEndpoint.kt [hunk 9]
```diff
@@ -33,13 +33,15 @@ class TestMessagingEndpoint : IntegrationTest() {
...
-        mockMvc.perform(requestMe)
+        val resp = mockMvc.perform(requestMe)
                 .andExpect(status().isOk)
                 .andExpect(content().contentType(APPLICATION_JSON_UTF_8))
                 .andExpect(jsonPath("$.id").exists())
                 .andExpect(jsonPath("$.groupMessageIds[0]").exists())
                 .andExpect(jsonPath("$.groupMessageIds[1]").doesNotExist())
                 .andReturn().response.contentAsString
+
+        println(resp)
```
Comment (florianschmidt1994) [inline src/test/java/backend/Integration/TestMessagingEndpoint.kt:44]: Why is this in here? Does it belong to this PR?

**Review by florianschmidt1994 — COMMENTED** [review]
I added some comments that we should address, looks pretty good besides that. Thanks!

File: src/main/java/backend/model/challenges/Challenge.kt (hunk 1 above)
Comment (florianschmidt1994) [reply src/main/java/backend/model/challenges/Challenge.kt:142]: We'll leave this for now, until we decided how to handle offline sponsors with this system now at all

Comment (florianschmidt1994) [reply src/main/java/backend/model/challenges/Challenge.kt:142]: https://github.com/BreakOutEvent/Documentation/issues/137

**Review by thepiwo — APPROVED** [review]

**Pull request merged**